activitysim/abm/models/tour_mode_choice_reallocation.py

```diff
@@ -0,0 +1,653 @@
+    assert len(logsum_trips) == len(trips) * len(tour_mode_alts)
+    logsum_trips.index.name = "trip_id"
+
+    return logsum_trips
+
+
+def append_tour_leg_trip_mode_choice_logsums(tours):
+    """
+    Creates trip mode choice logsum column in tours table for each tour mode and leg
+
+    Parameters
+    ----------
+    tours : pd.DataFrame
+
+    Returns
+    -------
+    tours : pd.DataFrame
+        Adds two * n_modes logsum columns to each tour row, e.g. "logsum_DRIVE_outbound"
+    """
+    trips = inject.get_table("trips").to_frame()
+    trip_dir_mode_logsums = trips.pivot(
+        index="tour_id",
+        columns=["tour_mode", "outbound"],
+        values="trip_mode_choice_logsum",
+    )
+    new_cols = [
+        "_".join(["logsum", mode, "outbound" if outbound else "inbound"])
+        for mode, outbound in trip_dir_mode_logsums.columns
+    ]
+    trip_dir_mode_logsums.columns = new_cols
+    trip_dir_mode_logsums.reindex(tours.index)
+    tours = pd.merge(tours, trip_dir_mode_logsums, left_index=True, right_index=True)
+
+    return tours
+
+
+def get_trip_mc_logsums_for_all_modes(tours, segment_column_name, model_settings, trace_label):
+    """
+    Creates pseudo-trips from tours and runs trip mode choice to get logsums
+
+    Parameters
+    ----------
+    tours : pandas.DataFrame
+    segment_column_name : str
+        column in tours table used for segmenting model spec
+    model_settings : dict
+    trace_label : str
+
+    Returns
+    -------
+    tours : pd.DataFrame
+        Adds two * n_modes logsum columns to each tour row, e.g. "logsum_DRIVE_outbound"
+    """
+
+    # create pseudo-trips from tours for all tour modes
+    logsum_trips = create_logsum_trips(
+        tours, segment_column_name, model_settings, trace_label
+    )
+
+    # temporarily register trips in the pipeline
+    pipeline.replace_table("trips", logsum_trips)
+    tracing.register_traceable_table("trips", logsum_trips)
+    pipeline.get_rn_generator().add_channel("trips", logsum_trips)
+
+    # run trip mode choice on pseudo-trips. use orca instead of pipeline to
+    # execute the step because pipeline can only handle one open step at a time
+    orca.run(["trip_mode_choice"])
+
+    # add trip mode choice logsums as new cols in tours
+    tours = append_tour_leg_trip_mode_choice_logsums(tours)
+
+    # de-register logsum trips table
+    pipeline.get_rn_generator().drop_channel("trips")
+    tracing.deregister_traceable_table("trips")
+
+    return tours
+
+
+def code_household_auto_sufficiency(households):
+    """
+    Adds and codes household auto sufficiency based on household auto ownership and number of drivers in the household
+
+    Parameters
+    ----------
+    households : pd.Dataframe
+
+    Returns
+    -------
+    households : pd.Dataframe
+        Household table with auto sufficiency column added
+
+    """
+    households['auto_sufficiency'] = 'suff'
```

Comment:
@TianYangLin Does this column not exist in the households table by the time the model gets to tour mode choice? Other parts of the code seems to indicate the presence of a `car_sufficiency` column.

Reply:
There are references to a "car_sufficiency" in a commented out section of joint_tour_participation and in expressions in some csvs, but is not found used in the model.